samza-core/src/main/java/org/apache/samza/storage/TaskSideInputHandler.java

```diff
@@ -0,0 +1,331 @@
+      Map<SystemStreamPartition, Object> sspLockObjects,
+      Map<SystemStreamPartition, Optional<String>> checkpointOffsets) {
+    this.taskName = taskName;
+    this.taskSideInputStorageManager = taskSideInputStorageManager;
+    this.systemAdmins = systemAdmins;
+    this.streamMetadataCache = streamMetadataCache;
+    this.sideInputTasksCaughtUp = sideInputTasksCaughtUp;
+    this.storeToProcessor = storeToProcessor;
+    this.sspLockObjects = Collections.unmodifiableMap(sspLockObjects);
+    this.checkpointedOffsets = Collections.unmodifiableMap(checkpointOffsets);
+
+    this.sspToStores = storeToSSPs.entrySet().stream()
+        .flatMap(storeAndSSPs -> storeAndSSPs.getValue().stream()
+            .map(ssp -> new AbstractMap.SimpleImmutableEntry<>(ssp, storeAndSSPs.getKey())))
+        .collect(Collectors.groupingBy(
+            Map.Entry::getKey,
+            Collectors.mapping(Map.Entry::getValue, Collectors.toSet())));
+
+    // for non-checkpointed SSPs, use their newest offset. for checkpointed SSPs, use their current checkpoint
+    this.sspBootstrapOffsets = new HashMap<>();
+    initialSSPMetadata.entrySet().stream()
+        // only SSPs for this task
+        .filter(entry -> this.sspToStores.containsKey(entry.getKey()))
+        // that do not have checkpoints
+        .filter(entry -> !this.checkpointedOffsets.containsKey(entry.getKey()))
+        .forEach(entry -> this.sspBootstrapOffsets.put(entry.getKey(), entry.getValue().getNewestOffset()));
+    this.checkpointedOffsets.entrySet().stream()
+        .filter(entry -> entry.getValue().isPresent())
+        .forEach(entry -> this.sspBootstrapOffsets.put(entry.getKey(), entry.getValue().get()));
+
+    this.lastProcessedOffsets = taskSideInputStorageManager.getFileOffsets();
+    this.startingOffsets = getStartingOffsets(this.lastProcessedOffsets, getOldestOffsets());
+    this.startingOffsets.forEach((ssp, offset) -> checkCaughtUp(ssp, offset, true));
+
+
+    Set<String> checkpointedStores = this.sspToStores.entrySet().stream()
+        .filter(sspAndStores -> this.checkpointedOffsets.containsKey(sspAndStores.getKey()))
+        .flatMap(sspAndStores -> sspAndStores.getValue().stream())
+        .collect(Collectors.toSet());
+
+    Set<String> nonCheckpointedStores = this.sspToStores.entrySet().stream()
+        .filter(sspAndStores -> !this.checkpointedOffsets.containsKey(sspAndStores.getKey()))
+        .flatMap(sspAndStores -> sspAndStores.getValue().stream())
+        .collect(Collectors.toSet());
+
+    this.checkpointedSSPExecutor = new KeyBasedExecutorService(Math.max(1, checkpointedStores.size()));
+    this.nonCheckpointedSSPExecutor = new KeyBasedExecutorService(Math.max(1, nonCheckpointedStores.size()));
+  }
+
+  public void process(IncomingMessageEnvelope envelope, TaskCallbackFactory callbackFactory) {
+    TaskCallback callback = callbackFactory.createCallback();
```

Comment:
The reason we have callback factory at the TaskInstance layer is to have more precision around the triggering of timeout for callback and have it start right before dispatching the message to the user logic. If we are considering SideInputHandler similar to user logic then we should minimize the exposure of factory to the handler; If not, we should move this closer to the `SideInputProcessor` to reflect the semantics of callback timeout.

Looks like we may need to move this closer to the process as we can get blocked on the lock which could trigger callback timeout.

Reply:
I don't think we can move this all the way down to `SideInputsProcessor`, because for some SSPs, for a message to be considered "processed", a number of `SideInputsProcessor`s will need to be invoked. So the callback factory needs to be a scope higher than that.

Timeout is a problem though, and I think we do need to move this down at least to after the block waiting on the lock.

What do you think?